tools/build_langserver/langserver/rename.go

```diff
@@ -0,0 +1,164 @@
+package langserver
+
+import (
+	"context"
+	"core"
+	"encoding/json"
+	"parse/asp"
+	"tools/build_langserver/lsp"
+
+	"github.com/sourcegraph/jsonrpc2"
+)
+
+const renameMethod = "textDocument/rename"
+
+func (h *LsHandler) handleRename(ctx context.Context, req *jsonrpc2.Request) (result interface{}, err error) {
+	if req.Params == nil {
+		return nil, nil
+	}
+
+	var params lsp.RenameParams
+	if err := json.Unmarshal(*req.Params, &params); err != nil {
+		return nil, err
+	}
+
+	documentURI, err := getURIAndHandleErrors(params.TextDocument.URI, renameMethod)
+	if err != nil {
+		return nil, err
+	}
+
+	h.mu.Lock()
+	defer h.mu.Unlock()
```

Comment:
careful holding the mutex around a big function like `getRenameEdits`. Not saying it's wrong but locking for shorter periods of time might be more performant & help avoid deadlocks.